### ISSUE TITLE
Upgrade flake8 in CI and pre-commit hook.

### ISSUE THREAD

**Issue by MortalHappiness** [ray-project/ray#48022]
This is a subtask of https://github.com/ray-project/ray/issues/47991. See the parent issue for more information.

**Comment by MortalHappiness** [comment]
@CheyuWu could you tell me which rules should be ignored? I'll open seperate sub-issues for them

**Comment by CheyuWu** [comment]
Since the latest flake8 dependency requires modifying `requirements_compiled.txt` for `mccabe` and `pyflakes`, I need to add `requirements_compiled.txt` to commit.

**Comment by CheyuWu** [comment]
- **B018** Found useless Tuple expression. Consider either assigning it to a variable or removing it.
  - **B019** Use of `functools.lru_cache` or `functools.cache` on methods can lead to memory leaks. The cache may retain instance references, preventing garbage collection.
  - **B020** Loop control variable overrides iterable it iterates
  - **B023** unctions defined inside a loop must not use variables redefined in the loop, because [late-binding closures are a classic gotcha](https://docs.python-guide.org/writing/gotchas/#late-binding-closures).
  - **B024** Abstract base class has methods, but none of them are abstract. This is not necessarily an error, but you might have forgotten to add the @abstractmethod decorator, potentially in conjunction with @classmethod, @property and/or @staticmethod.
  - **B026** Star-arg unpacking after a keyword argument is strongly discouraged, because it only works when the keyword parameter is declared after all parameters supplied by the unpacked sequence, and this change of ordering can surprise and mislead readers. There was https://github.com/python/cpython/issues/82741, but legacy usage and parser limitations make it difficult.
  - **B027** Empty method in abstract base class, but has no abstract decorator. Consider adding @abstractmethod.
  - **B028** No explicit stacklevel argument found. The warn method from the warnings module uses a stacklevel of 1 by default. This will only show a stack trace for the line on which the warn method is called. It is therefore recommended to use a stacklevel of 2 or greater to provide more information to the user.
  - **B033** Sets should not contain duplicate items. Duplicate items will be replaced with a single item at runtime.
  - **B035** Found dict comprehension with a static key - either a constant value or variable not from the comprehension expression. This will result in a dict with a single key that was repeatedly overwritten.
  - **B036** Found except BaseException: without re-raising (no raise in the top-level of the except block). This catches all kinds of things (Exception, SystemExit, KeyboardInterrupt...) and may prevent a program from exiting as expected.
  - **B039** `ContextVar` with mutable literal or function call as default. This is only evaluated once, and all subsequent calls to .get() would return the same instance of the default. This uses the same logic as B006 and B008, including ignoring values in extend-immutable-calls.
  - **E721** Do not compare types, use 'isinstance()
  - **E225** [Missing whitespace around operator](https://www.flake8rules.com/rules/E225.html)
  - **E231** [Missing whitespace after ',', ';', or ':'](https://www.flake8rules.com/rules/E231.html)
  - **E501** [Line too long (82 > 79 characters)](https://www.flake8rules.com/rules/E501.html)
  - **F401** [Module imported but unused](https://www.flake8rules.com/rules/F401.html)
  - **F811** [Redefinition of unused name from line n](https://www.flake8rules.com/rules/F811.html)

**Comment by aslonnie** [comment]
@CheyuWu and @MortalHappiness , do you two know each other and work together?

**Comment by MortalHappiness** [comment]
@aslonnie Recently, some friends of mine told me that they want to try contributing to open-source projects, so I created some issues in Ray and KubeRay for them to work on. @CheyuWu is one of them.

**Comment by MortalHappiness** [comment]
@anyscalesam I have upgraded a linter before, and from my experience, it’s better to upgrade the linter first and ignore the new rules for now. This is because linter upgrades usually take some time, and new commits will likely be added during that period. If you only update the linter configuration at the end, you might have to fix new style errors caused by the latest commits. That’s why I set the first subtask to upgrade the linter but ignore the new rules for now.
